bindings/java/src/main/java/org/ucx/jucx/LoadLibrary.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2017.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+package org.ucx.jucx;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import org.slf4j.Logger;
+import org.ucx.jucx.util.JucxLogger;
+import org.ucx.jucx.util.JucxLogger.LoggerType;
+
+public class LoadLibrary {
+    private static final Logger logger = JucxLogger.getLogger(LoggerType.API);
+
+    /**
+     * Tries to load the library, by extracting it from the current class jar.
+     * If that fails, falls back on {@link System#loadLibrary(String)}.
+     *
+     * @param String
+     *            to the library name to be extracted and loaded from the this
+     *            current jar
+     * @return true if the the file loaded successfully, or false on error
+     */
+    public static boolean loadLibrary(String resourceName) {
+        ClassLoader loader = LoadLibrary.class.getClassLoader();
+
+        // Search shared object on java classpath
+        URL url = loader.getResource(resourceName);
+        File file = null;
+        try { // Extract shared object's content to a generated temp file
+            file = extractResource(url);
+        } catch (IOException e) {
+            logger.error("Native code library failed to extract URL: "
+                        + url, e);
+            return false;
+        }
+
+        if (file != null && file.exists()) {
+            String filename = file.getAbsolutePath();
+            try { // Load shared object to JVM
+                System.load(filename);
+            } catch (UnsatisfiedLinkError e) {
+                logger.error("Native code library failed to load: "
+                            + resourceName, e);
+                return false;
+            }
+            // Delete temp file
+            if (file.delete() == false) {
```

Comment:
should we also delete the temp dir?

Reply:
It looks like the `tempDir` just gets created once, then persists between calls to loadLibrary creating new temp files in the directory. I suppose it doesn't sounds too bad to create and delete the temp directory five times. Though this would mean `extractResource` needs to return the directory as well as the file. Or assume the file exists in a directory to be deleted. Using the class static `tempDir` as something to delete at the end of `LoadLibrary` isn't threadsafe, though that might not matter.

Reply:
@gmegan How about [File.deleteOnExit()](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#deleteOnExit())?
From the doc: "Files (or directories) are deleted in the reverse order that they are registered.". So I can register them in the right order and when the Java process terminates - the files and the directory will be deleted.